component/discovery/file/file.go

```diff
@@ -0,0 +1,316 @@
+package file
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/grafana/agent/component/discovery"
+
+	"github.com/bmatcuk/doublestar"
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "discovery.file",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the discovery.file
+// component.
+type Arguments struct {
+	Paths         []string      `river:"paths,attr"`
+	ExcludedPaths []string      `river:"excluded_paths,attr,optional"`
+	UpdatePeriod  time.Duration `river:"update_period,attr,optional"`
+}
```

Comment:
As is, this doesn't enable prevent Kubernetes-focused use cases, which IMO are a requirement for the initial release of these components. 

For collecting logs from Kubernetes pods which are mounted to the host, the expectation is that you would chain `discovery.kubernetes` to `discovery.relabel` to `discovery.file` to `loki.source.file` to be able to discover log files from containers and decorate it with metadata about the pods, something like:

```river 
discovery.kubernetes "example" {
  role = "pod"
}

discovery.relabel "example" {
  targets = [discovery.kubernetes.example.targets] 
 
  rule {
    source_labels = ["__meta_kubernetes_namespace", "__meta_kubernetes_pod_label_name"]
    target_label  = "job"
    separator     = "/"
  }

  rule {
    source_labels = ["__meta_kubernetes_pod_uid", "__meta_kubernetes_pod_container_name"]
    target_label  = "__path__"
    separator     = "/" 
    replacement   = "/var/log/pods/*$1/*.log"
  } 
}

discovery.file "example" {
  targets = [discovery.relabel.example.output] 
}

loki.source.file "example" {
  targets    = [discovery.file.example.targets] 
  forward_to = [...]
}
```

This is the equivalent of the following Promtail config file:

```yaml 
scrape_config:
- job_name: kubernetes-pods 
  kubernetes_sd_configs:
  - role: pod 
  relabel_configs:
  - source_labels:
    - __meta_kubernetes_namespace 
    - __meta_kubernetes_pod_label_name 
    target_label: job 
    separator: /
  - source_labels: 
    - __meta_kubernetes_pod_uid
    - __meta_kubernetes_pod_container_name
    target_label: __path__ 
    separator: / 
    replacement: /var/log/pods/*$1/*.log
```

I know we're planning on exploring retrieving logs directly from the Kubernetes API server with #2504, but it won't ship with v0.30 and it's not guaranteed it'll even work at scale. At least in the interrim, we need to allow `discovery.file` to be used in a pipeline where targets are discovered from Kubernetes.

Reply:
Good point and resolved to take in targets.